### PR TITLE
SONAR-6990 add and use a duplication repository in ce

### PR DESCRIPTION
**Pull request by sns-seb**
this PR contains:
* implementation of a Duplication Repository and replacement of every existing access to duplication information from the batch bu use of this new repository
* some UT refactoring
    * move `ComponentProvider` and provide a noop implementation to avoid null properties in Rules making use of it
* some CE refactoring
    * merged `ReportTreeRooHolder` into `TreeRootHolder` and add new methods to it (to seek components by their key)

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistDuplicationsStep.java [hunk 1]
```diff
@@ -120,23 +119,30 @@ private String createXmlDuplications(String componentKey, Iterator<BatchReport.D
...
         appendDuplication(xml, componentKey, duplicate);
+      } else if (duplicate instanceof InProjectDuplicate) {
+        // Duplication is on a different file
+        appendDuplication(xml, ((InProjectDuplicate) duplicate).getFile().getKey(), duplicate);
+      } else if (duplicate instanceof CrossProjectDuplicate) {
+        // componentKey is only set for cross project duplications
+        String crossProjectComponentKey = ((CrossProjectDuplicate) duplicate).getFileKey();
+        appendDuplication(xml, crossProjectComponentKey, duplicate);
+      }
+      else {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistDuplicationsStep.java]: MINOR Move this "else" on the same line that the previous closing curly brace. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ARightCurlyBraceSameLineAsNextBlockCheck)

File: server/sonar-server/src/main/java/org/sonar/server/computation/source/DuplicationLineReader.java [hunk 2]
```diff
@@ -20,93 +20,121 @@
...
+    return from(duplicatedTextBlocks)
+        .toMap(new Function<TextBlock, Integer>() {
+          int i = 1;
 
-  private static boolean matchLine(BatchReport.TextRange range, int line) {
-    return range.getStartLine() <= line && line <= range.getEndLine();
+          @Nullable
+          @Override
+          public Integer apply(TextBlock input) {
+            return i++;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/source/DuplicationLineReader.java]: CRITICAL Remove this increment or correct the code not to waste it. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2123)

File: server/sonar-server/src/main/java/org/sonar/server/computation/source/DuplicationLineReader.java [hunk 3]
```diff
@@ -20,93 +20,121 @@
...
-    int line = lineBuilder.getLine();
-    List<BatchReport.TextRange> blocks = findDuplicationBlockMatchingLine(line);
-    for (BatchReport.TextRange block : blocks) {
-      lineBuilder.addDuplication(duplicationIdsByRange.get(block));
+    Predicate<Map.Entry<TextBlock, Integer>> containsLine = new TextBlockContainsLine(lineBuilder.getLine());
+    for (Integer textBlockIndex : from(duplicatedTextBlockIndexByTextBlock.entrySet())
+        .filter(containsLine)
+        .transform(MapEntryToBlockId.INSTANCE)
+            // list is sorted to cope with the non-guaranteed order of Map entries which would trigger false detection of changes
+            // in {@link DbFileSources.Line#getDuplicationList()}
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/source/DuplicationLineReader.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: server/sonar-server/src/main/java/org/sonar/server/computation/source/DuplicationLineReader.java [hunk 4]
```diff
@@ -20,93 +20,121 @@
...
+   * <p>
+   * This method uses the natural order of TextBlocks to ensure that given the same set of TextBlocks, they get the same
+   * index. It avoids false detections of changes in {@link DbFileSources.Line#getDuplicationList()}.
+   * </p>
+   */
+  private static Map<TextBlock, Integer> createIndexOfDuplicatedTextBlocks(Collection<Duplication> duplications) {
+    List<TextBlock> duplicatedTextBlocks = extractAllDuplicatedTextBlocks(duplications);
+    Collections.sort(duplicatedTextBlocks);
+    return from(duplicatedTextBlocks)
+        .toMap(new Function<TextBlock, Integer>() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/source/DuplicationLineReader.java]: MAJOR Make this a named "static" inner class. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2694)

File: server/sonar-server/src/main/java/org/sonar/server/computation/duplication/Duplication.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.FluentIterable.from;
+import static java.util.Objects.requireNonNull;
+
+@Immutable
+public final class Duplication {
+  private static Ordering<Duplicate> DUPLICATE_ORDERING = Ordering.from(DuplicateComparatorByType.INSTANCE)
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/duplication/Duplication.java]: MINOR Rename this field "DUPLICATE_ORDERING" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS3008)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* CRITICAL 1 critical
* MAJOR 2 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

**Comment by julienlancelot** [comment]
LGTM

**Comment by sns-seb** [comment]
feedback applied, waiting for IT build before merging

File: server/sonar-server/src/main/java/org/sonar/server/computation/source/DuplicationLineReader.java [hunk 6]
```diff
@@ -20,93 +20,113 @@
...
     }
   }
 
+  private static class TextBlockIndexGenerator implements Function<TextBlock, Integer> {
+    int i = 1;
+
+    @Nullable
+    @Override
+    public Integer apply(TextBlock input) {
+      return i++;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/source/DuplicationLineReader.java]: CRITICAL Remove this increment or correct the code not to waste it. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2123)

File: server/sonar-server/src/main/java/org/sonar/server/computation/source/DuplicationLineReader.java [hunk 7]
```diff
@@ -20,93 +20,113 @@
...
-    int line = lineBuilder.getLine();
-    List<BatchReport.TextRange> blocks = findDuplicationBlockMatchingLine(line);
-    for (BatchReport.TextRange block : blocks) {
-      lineBuilder.addDuplication(duplicationIdsByRange.get(block));
+    Predicate<Map.Entry<TextBlock, Integer>> containsLine = new TextBlockContainsLine(lineBuilder.getLine());
+    for (Integer textBlockIndex : from(duplicatedTextBlockIndexByTextBlock.entrySet())
+      .filter(containsLine)
+      .transform(MapEntryToBlockId.INSTANCE)
+      // list is sorted to cope with the non-guaranteed order of Map entries which would trigger false detection of changes
+      // in {@link DbFileSources.Line#getDuplicationList()}
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/source/DuplicationLineReader.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* CRITICAL 1 critical
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Pull request merged**